### PR TITLE
[CALCITE-3285] EnumerableMergeJoin should support non-equi join conditions

### PR DESCRIPTION
**Pull request by rubenada**
Jira ticket: [CALCITE-3285](https://issues.apache.org/jira/browse/CALCITE-3285).
Added an extra predicate for non-equi condition in EnumerableDefaults#MergeJoinEnumerator (it will be null in case of equi-joins).

### CODE REVIEW

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java [hunk 1]
```diff
@@ -3911,9 +3936,33 @@ private boolean advance() {
...
+            }
+          }
+        }
+        if (results.isEmpty()) {
+          if (done) {
+            return false;
+          }
+          return advance();
+        } else {
+          cartesians = Linq4j.enumerator(results);
```
Comment (vlsi) [inline linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: How is this different from a filter on top of `Linq4j.product`?

If it is not different, then please remove manual for-for.
If it is different please add a clarification comment.

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java [hunk 2]
```diff
@@ -1971,6 +1972,26 @@ private void closeInner() {
...
+   * Inputs must sorted in ascending order, nulls last.
+   * NOTE: The current API is experimental and subject to change without notice.
+   */
+  @Experimental
+  public static <TSource, TInner, TKey extends Comparable<TKey>, TResult> Enumerable<TResult>
+      mergeJoin(final Enumerable<TSource> outer,
+      final Enumerable<TInner> inner,
+      final Function1<TSource, TKey> outerKeySelector,
+      final Function1<TInner, TKey> innerKeySelector,
+      final Predicate2<TSource, TInner> extraPredicate,
```
Comment (vlsi) [inline linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Please add the relevant documentation that clarifies why this `extraPredicate` is needed and how it is different from a predicate on top of join results.

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java [hunk 3]
```diff
@@ -1971,6 +1972,26 @@ private void closeInner() {
...
+        resultSelector, generateNullsOnLeft, generateNullsOnRight);
+  }
+
+  /**
+   * Joins two inputs that are sorted on the key, with an extra predicate for non equi-join
+   * conditions (in case of equi-join, it will be null).
+   * Inputs must sorted in ascending order, nulls last.
+   * NOTE: The current API is experimental and subject to change without notice.
+   */
+  @Experimental
```
Comment (vlsi) [inline linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: I would suggest `@API(...)`

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 1 above)
Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: At the current state (only INNER join supported), we could have a filter on top of the product. However, we have other join types in the backlog that can / will be supported by EnumerableMergeJoin (SEMI-join, see [CALCITE-3833](https://issues.apache.org/jira/browse/CALCITE-3833); and ANTI-join, see  [CALCITE-3834](https://issues.apache.org/jira/browse/CALCITE-3834)). For those join types (which do not project the right-hand-side of the join), we cannot use the strategy of applying a filter on top of the product, so I am just preparing the ground here with an homogeneous strategy that can work on other join types.

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 2 above)
Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: You're right, I'll document that we use a predicate rather that a filter on top of join results in order to support other join types in the (near) future.

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 3 above)
Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Ok, I'll change it.

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 1 above)
Comment (vlsi) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Implementation-wise it looks the same.
I see no reason why do you use `for-for-if` here.

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 3 above)
Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Unfortunately, when I try to import `org.apiguardian.api.API;` and use  `@API(since = "1.23", status = API.Status.EXPERIMENTAL)` in EnumerableDefaults (which belongs to linq4j), I get a compilation error. Is it possible to use that library there? Do I need to change gradle config files to do so?

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 1 above)
Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: The next step will be implementing SEMI join type with Enumerable Merge Join. In order to do so, we would just a need a `break` inside the `if` (to avoid 'left' duplicates, which SEMI join must not have). So this algorithm works for INNER join and can be easily adapted for other types, like SEMI.

Comment (vlsi) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Is it different from the existing `org.apache.calcite.linq4j.EnumerableDefaults#semiJoin` and `#antiJoin` methods?

Why do you reimplement the same thing?

Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: `org.apache.calcite.linq4j.EnumerableDefaults#semiJoin` and `#antiJoin` methods implement hashSemiJoin and hashAntiJoin, they are used by EnumerableHashJoin, the idea is to provide another alternative implementation of semiJoin/antiJoin via MergeJoin algorithm (as they exist for NestedLoopJoin, CorrelateJoin or CorrelateBatchJoin)

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 3 above)
Comment (vlsi) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: It comes from `apiguardian`, so you need to add the relevant dependency to `linq4j`: https://github.com/apache/calcite/blob/d234626227954eefffe49f42abec65c649ffe3a6/core/build.gradle.kts#L44

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 1 above)
Comment (vlsi) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Let me rephrase it differently:

1) You calculate `lefts` and `rights` using merge join algorithm on `equal` predicates
2) Then you have `lefts` and `rights`, and there's nothing from `merge` join left that you can do about it. The rest could be exactly the same as in `hash semi-join`.

`for-for-if` logic performs exactly the same actions as the existing `semiJoin` and `antoJoin` methods. Why duplicate code (bugs)?

Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Ok, I think I understand your point. We should re-use existing code instead of the `for-for-if`. However, I don't think hashJoin code is the appropriate module here, since there is no need to create the hash structure to find matching keys (because we already have them with the merge join algorithm). Instead, we should use the `nestedLoopJoin(lefts, rights, extraPredicateWithNonEquiClauses)` which performs precisely this `for-for-if` process.

Comment (vlsi) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: > I don't think hashJoin code is the appropriate module here, since there is no need to create the hash structure to find matching keys

Well, my point is the code seems to duplicate the existing code, so it looks like a bug.
Of course, the only predicates left are `non-equality`, so there's nothing to hash. 

However, the key point here is to avoid code duplication.

Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Refactored to re-use nested loop join

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 3 above)
Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: added apiguardian

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 2 above)
Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: documentation updated

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java [hunk 4]
```diff
@@ -3951,6 +3982,24 @@ public void close() {
...
+    private final Function2<TOuter, TInner, TResult> resultSelector;
+
+    CartesianProductJoinEnumerator(Function2<TOuter, TInner, TResult> resultSelector,
+                                   Enumerator<Object> outer, Enumerator<Object> inner) {
+      super(ImmutableList.of(outer, inner));
+      this.resultSelector = resultSelector;
+    }
+
+    @Override public TResult current() {
+      final List<Object> list = Arrays.asList(elements.clone());
```
Comment (vlsi) [inline linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: What is the purpose of `.clone()` here?

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java [hunk 5]
```diff
@@ -3951,6 +3982,24 @@ public void close() {
     }
   }
 
+  private static class CartesianProductJoinEnumerator<TResult, TOuter, TInner>
+      extends CartesianProductEnumerator<Object, TResult> {
+    private final Function2<TOuter, TInner, TResult> resultSelector;
+
+    CartesianProductJoinEnumerator(Function2<TOuter, TInner, TResult> resultSelector,
+                                   Enumerator<Object> outer, Enumerator<Object> inner) {
```
Comment (vlsi) [inline linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Should it be typed like `Enumerator<TOuter>`?

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 4 above)
Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Transforming `elements` array into list. Just applying the same approach as `Linq4j#CartesianProductListEnumerator`

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 5 above)
Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Not possible if we want to extend `CartesianProductEnumerator`:
```
/**
 * Enumerator over the cartesian product of enumerators.
 * @param <T> Input element type
 * @param <E> Element type
 */
public abstract class CartesianProductEnumerator<T, E> implements Enumerator<E> {...}
```

Comment (vlsi) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Well, can you cast when calling the constructor?

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 4 above)
Comment (vlsi) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Note: `list` never escapes this method, so there's no need to copy it.
There's no need to create `list` as well.

Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: you're right, I'll change it

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 5 above)
Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Modified: cast in constructor

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 4 above)
Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: modified: list no longer generated

**Comment by rubenada** [comment]
Thanks for your feedback @vlsi !
Do you think the PR is in a good shape in order to be merged?

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 2 above)
Comment (vlsi) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Thanks. However, can you please add `@param extraPredicate` that would clarify the meaning of `extraPredicate` in a couple of words?

The documentation does not need to clarfy all the implementation details, however, the naming of `Predicate2<TSource, TInner> extraPredicate` provides absolutely no clue one meaning of the predicate.

Well, I do see it is a predicate. I can judge that by class name. As a user of that API I want to know what is expected to be passed there. I want to know if `nulls` are allowed or not and so on.

Comment (rubenada) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: javadoc updated: added `@param extraPredicate`

**Comment by rubenada** [comment]
@vlsi I have carried out all your suggested modifications, do you think we could merge the patch?

**Comment by vlsi** [comment]
Thanks, it looks good to me.

**Comment by rubenada** [comment]
Thanks @vlsi , your suggestions have improved the patch very much.
